### PR TITLE
Add doc for `assert_queries` and `assert_no_queries`

### PR DESCRIPTION
**Pull request by yykamei**
Follow-up to #50281

### Motivation / Background

I think `assert_queries` and `assert_no_queries` are so useful that I want to read the documentation for the assertions. This pull request adds the doc for them and updated the test case named `test_assert_queries` to verify `:matcher` works as expected.

### Additional information

I also changed `assert_queries` to make sure the subscriber of `"sql.active_record"` is unsubscribed without affecting other test cases. Without this change, some test cases fail because of execution of `String#match`, which may cause `ArgumentError: invalid byte sequence in UTF-8`.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
    - This does not include behavior changes

### CODE REVIEW

**Review by skipkayhil — APPROVED** [review]
Looks good to me, will need another review from a committer :+1:

**Comment by yahonda** [comment]
`QueryLogsTest#test_invalid_encoding_query` test gets `ArgumentError: invalid byte sequence in UTF-8` for most of Ruby versions and database adapters. I understand this commit does not change the `QueryLogsTest#test_invalid_encoding_query` test itself, but not sure if they are not related.

https://buildkite.com/rails/rails/builds/102711#018c5c47-a1c3-46da-aea1-457b9f7d70bc/1096-1105

**Comment by skipkayhil** [comment]
> I understand this commit does not change the `QueryLogsTest#test_invalid_encoding_query` test itself, but not sure if they are not related.

Yeah this is really weird... the failures also seem distributed between different adapters on different version of Ruby. I'll retry a few to see if the QueryLogsTest is just flaky

File: activerecord/test/cases/assertions/query_assertions_test.rb
Comment (p8) [inline activerecord/test/cases/assertions/query_assertions_test.rb:30]: Thanks for adding this 🙏
Could you move this to a separate test case method?

**Comment by p8** [comment]
It seems QueryLogsTest is using these new assertions, which shouldn’t happen.

**Comment by skipkayhil** [comment]
> It seems QueryLogsTest is using these new assertions, which shouldn’t happen.

Ah, I think the subscriber stays active for all other tests once the QueryAssertions tests are run once, hence the flakiness.

**Comment by yahonda** [comment]
Would you squash your commits?

File: activerecord/test/cases/assertions/query_assertions_test.rb
Comment (yykamei) [reply activerecord/test/cases/assertions/query_assertions_test.rb]: Thank you! It makes sense to me. I moved the code with 7a268ecd1bd7296e0fea0a7663f812374da05087.

By the way, I couldn't run the test on my local machine. It's weird...

image

**Comment by yykamei** [comment]
Thank you. I squashed it, but I'm concerned about it. Is it OK to merge?

https://github.com/rails/rails/pull/50334#discussion_r1423488835

**Comment by yahonda** [comment]
Looks like this `end` is wrong.

https://github.com/rails/rails/blob/9c8f9643246d4d435c0a1cff64fe428de5140b32/activerecord/test/cases/assertions/query_assertions_test.rb#L24

**Comment by yahonda** [comment]
Opened a feature request at `rubocop-minitest` https://github.com/rubocop/rubocop-minitest/issues/279 to find tests that are not executed.

**Comment by yykamei** [comment]
Thank you! I understood the reason that the two cases were not run. I changed `assert_queries` a bit to make sure `ActiveSupport::Notifications.unsubscribe("sql.active_record")` is called in order to prevent other test cases from failing because of unexpected match operation.

**Comment by p8** [comment]
@yykamei It seems you just had bad luck with that CI run, the failure seems unrelated.
I'd just run the tests again.

**Comment by yykamei** [comment]
Thank you 🙏

**Pull request merged**
